### PR TITLE
gl_shader: free some GLSL shader data

### LINKED ISSUE #888: Memory leak when building shader permutation reported by heapusage
Reported by [heapusage](https://github.com/d99kris/heapusage), a valgrind-like tool (but much faster).

```
==434726== 109128 bytes in 359 block(s) are lost, originally allocated at:
==434726==    at 0x00007f4ee486e70f: calloc (humain.cpp:201)
==434726==    at 0x000055be96f99292: Z_TagMalloc(unsigned long, memtag_t) (qcommon.h:594)
==434726==    at 0x000055be96f9edff: CL_RefMalloc(int) (cl_main.cpp:2704)
==434726==    at 0x000055be97007596: GLShaderManager::UpdateShaderProgramUniformLocations(GLShader*, shaderProgram_t*) const (gl_shader.cpp:248)
==434726==    at 0x000055be9700b5dc: GLShaderManager::buildPermutation(GLShader*, int, int) (gl_shader.cpp:837)
==434726==    at 0x000055be9700b740: GLShaderManager::buildAll() (gl_shader.cpp:859)
==434726==    at 0x000055be970c122b: GLSL_FinishGPUShaders() (tr_shade.cpp:285)
==434726==    at 0x000055be97098164: RE_EndRegistration() (tr_init.cpp:1539)
==434726==    at 0x000055be96f43dd7: CL_InitCGame() (cl_cgame.cpp:640)
==434726==    at 0x000055be96f90003: CL_DownloadsComplete() (cl_download.cpp:111)
==434726==    at 0x000055be96f903db: CL_InitDownloads() (cl_download.cpp:238)
==434726==    at 0x000055be96faded5: CL_ParseGamestate(msg_t*) (cl_parse.cpp:462)
==434726==    at 0x000055be96fae1dd: CL_ParseServerMessage(msg_t*) (cl_parse.cpp:572)
==434726==    at 0x000055be96f9e6e9: CL_PacketEvent(netadr_t const&, msg_t*) (cl_main.cpp:2432)
==434726==    at 0x000055be96ec9280: Com_EventLoop() (common.cpp:431)
==434726==    at 0x000055be96eca235: Com_Frame() (common.cpp:1023)
==434726==    at 0x000055be96ec2e73: Application::ClientApplication::Frame() (ClientApplication.cpp:90)
==434726==    at 0x000055be97174831: Application::Frame() (Application.cpp:76)
==434726==    at 0x000055be971b474a: main (System.cpp:785)
```

### PR DESCRIPTION
**Pull request by illwieckz**
Fixes #888:

- https://github.com/DaemonEngine/Daemon/issues/888

### CODE REVIEW

File: src/engine/renderer/gl_shader.cpp [hunk 1]
```diff
@@ -219,6 +219,27 @@ GLShaderManager::~GLShaderManager()
 
 void GLShaderManager::freeAll()
 {
+	for ( GLShader* &shader : _shaders )
+	{
+		shader->_vertexShaderText.clear();
```
Comment (DolceTriade) [inline src/engine/renderer/gl_shader.cpp:224]: Why is this required? Shouldn't this be properly free'd by the std::string deconstructor?

Comment (illwieckz) [reply src/engine/renderer/gl_shader.cpp:224]: I don't know, even calling `clear` without calling `shrink_to_fit` doesn't stop the leak…

The leaked data seems to be the return of `GetShaderText()` in `BuildGPUShaderText()`.

Such string is a concatenation of multiple strings allocated with `reserve`, plus the return of `GetShaderText()`. That's all I know.

**Comment by slipher** [comment]
#892

**Comment by illwieckz** [comment]
@slipher do https://github.com/DaemonEngine/Daemon/pull/892 takes care of all the issues this one tried to fix?

**Comment by slipher** [comment]
> @slipher do #892 takes care of all the issues this one tried to fix?

Yes and more :)

**Comment by illwieckz** [comment]
Great!
